firestore/tests/unit/test__helpers.py

```diff
@@ -86,6 +86,130 @@ def test___ne__type_differ(self):
+            field_path = self._make_one(part)
+            self.assertEqual(field_path.parts, (part, ))
+
+    def test_component(self):
+        field_path = self._make_one('a..b')
+        self.assertEquals(field_path.parts, ('a..b',))
+
+    def test_constructor_iterable(self):
+        field_path = self._make_one('a', 'b', 'c')
+        self.assertEqual(field_path.parts, ('a', 'b', 'c'))
+
+    def test_to_api_repr_a(self):
+        parts = 'a'
+        field_path = self._make_one(parts)
+        self.assertEqual('a', field_path.to_api_repr())
+
+    def test_to_api_repr_tick(self):
+        parts = '`'
+        field_path = self._make_one(parts)
+        self.assertEqual('`\``', field_path.to_api_repr())
+
+    def test_to_api_repr_slash(self):
+        parts = '\\'
+        field_path = self._make_one(parts)
+        self.assertEqual(r'`\\`', field_path.to_api_repr())
+
+    def test_to_api_repr_double_slash(self):
+        parts = r'\\'
+        field_path = self._make_one(parts)
+        self.assertEqual(r'`\\\\`', field_path.to_api_repr())
+
+    def test_to_api_repr_underscore_valid(self):
+        parts = '_33132'
+        field_path = self._make_one(parts)
+        self.assertEqual('_33132', field_path.to_api_repr())
+
+    def test_to_api_repr_number_invalid(self):
+        parts = '03'
+        field_path = self._make_one(parts)
+        self.assertEqual('`03`', field_path.to_api_repr())
+
+    def test_to_api_repr_valid_part(self):
+        parts = 'a0332432'
+        field_path = self._make_one(parts)
+        self.assertEqual('a0332432', field_path.to_api_repr())
+
+    def test_to_api_repr_chain(self):
+        parts = 'a', '`', '\\', '_3', '03', 'a03', '\\\\'
+        field_path = self._make_one(*parts)
+        self.assertEqual(r'a.`\``.`\\`._3.`03`.a03.`\\\\`',
+                         field_path.to_api_repr())
+
+    def test_from_string(self):
+        field_path = self._get_target_class().from_string('a.b.c')
+        self.assertEqual(field_path.parts, ('a', 'b', 'c'))
+
+    def test_list_splat(self):
+        parts = ['a', 'b', 'c']
+        field_path = self._make_one(*parts)
+        self.assertEqual(field_path.parts, ('a', 'b', 'c'))        
+
+    def test_tuple_splat(self):
+        parts = ('a', 'b', 'c')
+        field_path = self._make_one(*parts)
+        self.assertEqual(field_path.parts, ('a', 'b', 'c'))        
+
+    def test_invalid_chars_from_string_fails(self):
+        parts = '~*/[].'
+        for part in parts:
+            with self.assertRaises(ValueError):
+                field_path = self._get_target_class().from_string(part)
+
+    def test_list_fails(self):
+        parts = ['a', 'b', 'c']
+        with self.assertRaises(ValueError):
+            field_path = self._make_one(parts)            
+
+    def test_tuple_fails(self):
+        parts = ('a', 'b', 'c')
+        with self.assertRaises(ValueError):        
+            field_path = self._make_one(parts)
+
+    def test_key(self):
```

Comment:
This test is fine as it is, but it doesn't handle the case I was worried about. In `update` you'll accept a dict that can have either strings or FieldPaths as keys. You will then loop through the keys and call `FieldPath.from_string` on the strings. My concern is that someone can pass this to update:
```
{"a.b": 1, FieldPath("a", "b"): 2}
```
Those keys are not equal, but after you call `from_string` on the first they will denote the same field. I think you have to explicitly check for that in `update`.

Reply:
I will address this in the update code.